src/server/pkg/hashtree/hashtree.go

```diff
@@ -256,9 +277,12 @@ func (h *hashtree) PutFile(path string, blockRefs []*pfs.BlockRef) error {
 		}
 		insertStr(&node.DirNode.Children, child)
 		node.SubtreeSize += sizeGrowth
-		h.updateHash(parent)
+		h.changed[parent] = true
 		return nil
-	})
+	}); err != nil {
+		return err
+	}
+	return h.canonicalize("/")
```

Comment:
No you're right. I kept the interface the same in this PR so that I could refactor the code without changing the tests. Next PR changes tests, but the code change is smaller because a lot of it was done here.